daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/Infoset.scala

```diff
@@ -80,13 +83,13 @@ trait InfosetComplexElement extends InfosetElement {
 
 trait InfosetSimpleElement extends InfosetElement {
 
-  def dataValue: Any
+  def dataValue: DataValuePrimitiveNullable
```

Comment:
So is this whole use of the phantom types just so we can deal with the stupid sentinel value that indicates useNilForDefault?

Blech. I suggest some names then like adding "Hack" as a suffix as in DataValuePrimitiveNilDefaultHack instead of the seemingly purposeful DataValuePrimitiveNullable. 

Also, terminology suggestion. XML Schema uses the terms nil and nillable. Java and Scala use the term Null as in null pointer. We should be consistent with this. When we're talking about DFDL's notion of Nil we should use Nillable in the identifiers and type names. 

So I think i'm suggesting Nullable => Nillable throughout here.

Reply:
Now that I think about it, there isn't really a reason we could have NoValue=DataValue(null). I believe this is what we do with the Maybe class.

My understanding is the 'nullable type' is standard terminology for a type which has some form of a "no value" sentinal value. It is confusing here probably because almost all types in Scala are nullable (notably, AnyVal types are not nullable).

I would use Maybe,Optional, or Opt in the name, but those are already taken by our other answers to nullable types.

I have no major problem with DataValuePrimitiveOrNoValue, the type name is already pretty long as is.

Reply:
Disregard my first paragraph. NoValue=DataValue(null) is exactly what we are currently doing.

Reply:
Disregard my first paragraph. NoValue=DataValue(null) is exactly what we are currently doing.

Reply:
Disregard my first paragraph. NoValue=DataValue(null) is exactly what we are currently doing.

Reply:
If you want to keep Nullable as the name because "nullable type" is a standard term, then you need to explain that in comments, because it wasn't standard to me and I'm reasonably well informed on these things, though not expert.

Reply:
Comment added above the type alias